qiskit_dynamics/pulse/string_model_parser/operator_from_string.py

```diff
@@ -0,0 +1,174 @@
+    """
+
+    # construct single system operator
+    out = single_operator_from_string(op_label, subsystem_dims[subsystem_index])
+
+    # sort subsystem labels and dimensions
+    sorted_subsystem_keys, sorted_subsystem_dims = zip(
+        *sorted(zip(subsystem_dims.keys(), subsystem_dims.values()))
+    )
+
+    # get subsystem location in ordered list
+    subsystem_location = sorted_subsystem_keys.index(subsystem_index)
+
+    # tensor identity on right if subsystem_index is not first subsystem
+    if subsystem_location != 0:
+        total_dim = np.prod(sorted_subsystem_dims[:subsystem_location])
+        out = np.kron(out, ident(total_dim))
+
+    # tensor identity on left if subsystem_index not the last subsystem
+    if subsystem_location + 1 != len(sorted_subsystem_keys):
+        total_dim = np.prod(sorted_subsystem_dims[(subsystem_location + 1) :])
+        out = np.kron(ident(total_dim), out)
+
+    return out
+
+
+# functions for generating individual operators
+def a(dim: int) -> np.ndarray:
+    """Annihilation operator."""
+    return np.diag(np.sqrt(np.arange(1, dim, dtype=complex)), 1)
+
+
+def adag(dim: int) -> np.ndarray:
+    """Creation operator."""
+    return a(dim).conj().transpose()
+
+
+def N(dim: int) -> np.ndarray:
+    """Number operator."""
+    return np.diag(np.arange(dim, dtype=complex))
+
+
+def X(dim: int) -> np.ndarray:
+    """Generalized X operator, written in terms of raising and lowering operators."""
+    return a(dim) + adag(dim)
+
+
+def Y(dim: int) -> np.ndarray:
+    """Generalized Y operator, written in terms of raising and lowering operators."""
+    return -1j * (a(dim) - adag(dim))
+
+
+def Z(dim: int) -> np.ndarray:
+    """Generalized Z operator, written as id - 2 * N."""
+    return ident(dim) - 2 * N(dim)
+
+
+def ident(dim: int) -> np.ndarray:
+    """Identity operator."""
+    return np.eye(dim, dtype=complex)
+
+
+# operator names
+__operdict = {
+    "X": X,
+    "Y": Y,
+    "Z": Z,
+    "a": a,
+    "A": a,
+    "Sm": a,
+    "Sp": adag,
+    "C": adag,
+    "N": N,
+    "O": N,
+    "I": ident,
+}
+
+
+def single_operator_from_string(op_label: str, dim: int) -> np.ndarray:
+    """Generate a single operator from a string.
+
+    Helper function for operator_from_string, see its documentation for
+    label interpretation.
+
+    Args:
+        op_label: String representing operator.
+        dim: Dimension of operator.
+
+    Returns:
+        np.ndarray
+
+    Raises:
+        QiskitError: If op_label doesn't correspond to a known operator.
+    """
+
+    op_func = __operdict.get(op_label, None)
+    if op_func is None:
+        raise QiskitError("String {} does not correspond to a known operator.".format(op_label))
+
+    return op_func(dim)
+
+
+def dag(op: np.ndarray) -> np.ndarray:
+    """Apply dagger."""
+    return np.conjugate(np.transpose(op))
+
+
+# pylint: disable=invalid-name
+__funcdict = {"dag": dag}
+
+
+def apply_func(name: str, op: np.ndarray) -> np.ndarray:
```

Comment:
This function seems sus. You could just replace it in the parser code (which is the only spot it looks to be used) with something like 
```python
elif token.type in ["Func", "Ext"]:
     if token.name == "dag":
         stack.append(np.conjugate(np.transpose(stack.pop(-1)))
     else:
          stack.append(stack.pop(-1))
```
(possibly something even simpler if you know whats going on in that code block)

Reply:
yeah you're right - I think maybe it was like this in the original pulse simulator to enable further expansion of functionality later, but for now it's pointless. I've implemented your suggestion and removed `apply_func` and related things, though changed the else statement to raise an error.